docs/advanced-settings.md
# Advanced Settings

## Env variables

### Required

| Key       | Description                                                             | Example                                            |
|-----------|-------------------------------------------------------------------------|----------------------------------------------------|
| `TOKEN`   | Your bot token. Get it from [@BotFather]                                | `1234567890:A1BCd2EF3gH45IJK6lMN7oPqr8ST9UvWX0Yz0` |
| `MANAGER` | Telegram user ID(s) that can manage the bot. Get it from [@userinfobot] | `1234567890` or `1234567890;987654321` [^1]        |

### API settings

| Key               | Description                                                    | Example                                                             | Default      |
|-------------------|----------------------------------------------------------------|---------------------------------------------------------------------|--------------|
| `API_ID`          | [Your Telegram API ID][telegram_api]                           | `1025907`                                                           | (predefined) |
| `API_HASH`        | [Your Telegram API hash][telegram_api]                         | `452b0359b988148995f22ff0f4229750`                                  | (predefined) |
| `TELEGRAPH_TOKEN` | Telegraph API access token(s). Get [here][telegraph_api]. [^2] | `1a23b456c78de90f1a23b456c78de90f1a23b456c78de90f1a23b456c78d` [^1] |              |

[@BotFather]: https://t.me/BotFather

[@userinfobot]: https://t.me/userinfobot

[telegram_api]: https://core.telegram.org/api/obtaining_api_id

[telegraph_api]: https://api.telegra.ph/createAccount?short_name=RSStT&author_name=Generated%20by%20RSStT&author_url=https%3A%2F%2Fgithub.com%2FRongronggg9%2FRSS-to-Telegram-Bot

### Network settings

| Key                         | Description                                           | Example                        | Default                                             |
|-----------------------------|-------------------------------------------------------|--------------------------------|-----------------------------------------------------|
| `T_PROXY`                   | Proxy used to connect to the Telegram API [^3]        | `socks5://172.17.0.1:1080`     |                                                     |
| `R_PROXY`                   | Proxy used to fetch feeds [^3]                        | `socks5://172.17.0.1:1080`     |                                                     |
| `PROXY_BYPASS_PRIVATE`      | Bypass proxy for private IPs or not?                  | `1`                            | `0`                                                 |
| `PROXY_BYPASS_DOMAINS`      | Bypass proxy for listed domains                       | `example.com;example.net` [^1] |                                                     |
| `USER_AGENT`                | User-Agent                                            | `Mozilla/5.0`                  | `RSStT/$VERSION RSS Reader (+https://git.io/RSStT)` |
| `IPV6_PRIOR`                | Enforce fetching feeds over IPv6 firstly or not? [^4] | `1`                            | `0`                                                 |
| `VERIFY_TLS`                | Verify TLS certificate or not?                        | `0`                            | `1`                                                 |
| `TRAFFIC_SAVING`            | Enable network traffic saving mode or not? [^5]       | `1`                            | `0`                                                 |
| `LAZY_MEDIA_VALIDATION`     | Let Telegram DC to validate media or not? [^6]        | `1`                            | `0`                                                 |
| `HTTP_TIMEOUT`              | HTTP request timeout in seconds                       | `60`                           | `12`                                                |
| `HTTP_CONCURRENCY`          | HTTP request concurrency overall (0=unlimited)        | `0`                            | `1024`                                              |
| `HTTP_CONCURRENCY_PER_HOST` | HTTP request concurrency per host (0=unlimited)       | `0`                            | `16`                                                |
| `HTTP_MAX_LINE_SIZE`        | Maximum allowed size of lines in responses            | `32768`                        | `16384`                                             |
| `HTTP_MAX_FIELD_SIZE`       | Maximum allowed size of header fields in responses    | `32768`                        | `16384`                                             |
| `DNS_CACHE_TTL`             | TTL for the DNS cache in seconds (0=disable)          | `60`                           | `10`                                                |

### Misc settings

| Key                           | Description                                                                 | Example                                       | Default                                         |
|-------------------------------|-----------------------------------------------------------------------------|-----------------------------------------------|-------------------------------------------------|
| `ERROR_LOGGING_CHAT`          | Chat (user/channel/group) ID for error logging.                             | `-1001234567890`                              | The first user ID in `MANAGER`                  |
| `MULTIUSER`                   | Enable multi-user feature or not?                                           | `0`                                           | `1`                                             |
| `CRON_SECOND`                 | Run the feed monitoring task at the n-th second of each minute? (0-59)      | `30`                                          | `0`                                             |
| `IMG_RELAY_SERVER`            | Media relay server (https://github.com/Rongronggg9/rsstt-img-relay) URL     | `https://wsrv.nl/?url=`                       | `https://rsstt-img-relay.rongrong.workers.dev/` |
| `IMAGES_WESERV_NL`            | https://github.com/weserv/images instance                                   | `https://t0.nl/`                              | `https://wsrv.nl/`                              |
| `DATABASE_URL`                | Database URL [^7]                                                           | `postgres://user:pass@example.com:5432/table` | `sqlite:/path/to/config/db.sqlite3`             |
| `TABLE_TO_IMAGE`              | Convert tables to image (causing higher CPU load) or just drop them?        | `1`                                           | `0`                                             |
| `MANAGER_PRIVILEGED`          | Allow the bot manager to manipulate any users' subscriptions or not? [^8]   | `1`                                           | `0`                                             |
| `NO_UVLOOP`                   | Never enable `uvloop` (even if it is found) or not?                         | `1`                                           | `0`                                             |
| `MULTIPROCESSING`             | Enable multiprocessing (up to `min(3, CPU_COUNT)`) or not? [^9]             | `1`                                           | `0`                                             |
| `EXECUTOR_NICENESS_INCREMENT` | The niceness increment of subprocesses (if `MULTIPROCESSING=1`) and threads | `5`                                           | `2`                                             |
| `DEBUG`                       | Enable debug logging or not?                                                | `1`                                           | `0`                                             |

## Manager options

> Manager options are options stored in the database. The bot manager can change it by using the `/set_option` command.

| Key                          | Description                                                | Example                         | Default          |
|------------------------------|------------------------------------------------------------|---------------------------------|------------------|
| `default_interval`           | Default feed monitoring interval [^10]                     | `15`                            | `10`             |
| `minimal_interval`           | Minimal feed monitoring interval [^11] [^12]               | `10`                            | `5`              |
| `user_sub_limit`             | Subscription number limit for ordinary user [^13] [^12]    | `150`                           | `-1` (unlimited) |
| `channel_or_group_sub_limit` | Subscription number limit for channel or group [^13] [^12] | `150`                           | `-1` (unlimited) |
| `sub_limit_reached_message`  | Additional message attached to the limit reached warning   | `https://t.me/RSStT_Channel/58` |                  |

[^1]: Can be a list separated by `;`, `,`, `(space)`, `(linebreak)`, or `(tab)`.
[^2]: Refresh the page every time you get a new token. If you have a lot of subscriptions, make sure to get at least 5 tokens.
[^3]: If you would like to use a proxy in a docker container, but your proxy is in your host machine, the hostname should be `172.17.0.1` (Linux) or `host.docker.internal` (macOS/Windows). Note: your proxy program should also listen to it.
[^4]: Use with caution. Enabling it will enforce the bot to try to fetch feeds over IPv6 (fallback to IPv4 if failed), which may be helpful if your IPv4 address gets banned by some feed providers. If it is disabled (by default), the bot will still try to fetch feeds over IPv4 and IPv6, but there is no clear priority. You should firstly ensure that the bot has IPv6 connectivity, especially if run in docker.
[^5]: Use with caution. May cause media validating and sending slightly unreliable. Meanwhile, effectively disable webpage title detection for `<iframe>` tags, instead, the title will always be URL hostname.
[^6]: Use with caution. Help cut down network traffic further. If enabled, RSStT no longer fetches media and validates it. Effectively disable long-pic detection and partially disable icon detection.
[^7]: Ref: [https://tortoise-orm.readthedocs.io/en/latest/databases.html](). Note that Railway.app will automatically fill this env variable.
[^8]: Use with caution. If enabled, the bot manager can bypass the permission check before manipulating any users'/channels'/groups' subscriptions. The command format is like `/sub @username`, `/sub +9999999999` (ordinary user) or `/sub -1009999999999` (channel/group). Should only be used temporarily and be disabled after finishing the manipulation. This option is considered safe for bot users since the bot manager can always manipulate their subscriptions by manipulating the database manually.
[^9]: Only valid when there are more than 1 CPU core, otherwise the process count is always `1`. Enabling multiprocessing may help improve the performance on multicore CPUs if there are tons of subscriptions but consumes more memory. If your VPS comes with multiple cores but the performance of each is poor, you may want to enable this feature.
[^10]: After a user subscribes to a feed, the default monitoring interval is applied.
[^11]: The minimal monitoring interval a user can set for a subscription.
[^12]: The bot manager will not be limited by this value.
[^13]: Once reached the limit, no more subscriptions can be created. However, existing subscriptions will not be removed even if reaching the limit. As a bot manager, you can enable `MANAGER_PRIVILEGED` mode to manually unsubscribe their subscriptions.


CHANGELOG.md
# Changelog

## v1.6.1

### en

**This is a rushed release. It bumps the dependency `telethon` to the latest version. Please upgrade to this version
immediately to avoid being unable to login due to the outdated dependency.**

The bot is currently being actively developed on the `multiuser` branch, but has not been merged back yet to avoid
introducing breaking changes too early. If you would like to try the multi-user version, there is a public
demo [@RSStT_Bot](https://t.me/RSStT_Bot) .

#### New features

- `.env` file support (only for manual execution, not for docker)
- Unescape HTML-escaped post title
- Use the title as the content of a post if the latter is of no text

#### Changes

- Minor bugfixes
- Introduce some workarounds to avoid being flood-controlled frequently
- Introduce some deps to speedup HTTP requests

### zh-Hans

**这是一个仓促的发布。它将依赖 `telethon` 升级到了最新版本。请立即升级到这个版本以免由于依赖过时而无法登录。**

机器人正在 `multiuser` 分支上被活跃开发，但尚未被合并回来，以免过早引入重大变更。如果你想要尝试多用户版本，这里有一个公开的 demo [@RSStT_Bot](https://t.me/RSStT_Bot) 。

#### 新特性

- `.env` 文件支持 (仅在手动执行时支持，不支持 docker)
- 反转义受到 HTML 转义的文章标题
- 当文章内容不含有文本时，将标题作为文章的内容

#### 变更

- 一些小的错误修复
- 引入了一些变通解决方案以免频繁受到泛洪控制
- 引入了一些依赖以加速 HTTP 请求

## v1.6.0

### en

#### BREAKING CHANGE

- Telegram bot library has been migrated from `python-telegram-bot` (which uses HTTP Bot API and is synchronous)
  to `telethon` (which uses MTProto Bot API and is asynchronous)
    - However, to use MTProto Bot API, an API key is needed. The bot has 7 built-in API keys (collected from the
      Internet) and in most cases it should not be unable to log in. But if so, please obtain your own API key (
      see [docker-compose.yml.sample](https://github.com/Rongronggg9/RSS-to-Telegram-Bot/blob/53f11a473933e620d707c9d15f6d48737bd7a982/docker-compose.yml.sample#L43)
      for details)

#### New features

- Thanks to the migration of Telegram bot library, bot can now connect to its DC directly, need not detour through the
  HTTP Bot API and keep polling to get new messages. Which makes the bot receive and reply messages more rapidly and
  lightweightedly. Even if the HTTP Bot API is down, the bot can still run unaffectedly. (more
  details: [Advantages of MTProto over Bot API](https://docs.telethon.dev/en/latest/concepts/botapi-vs-mtproto.html#advantages-of-mtproto-over-bot-api)
  , [MTProto vs HTTP Bot API](https://github.com/LonamiWebs/Telethon/wiki/MTProto-vs-HTTP-Bot-API))
- Support parsing more HTML elements
    - `<iframe>`
    - `<video><source><source>...</video>`
    - `<code>`
    - `<pre>`
- Support OPML importing and exporting
- Support sending too-long post via Telegraph (env var `TELEGRAPH_TOKEN` must be set)
- Support redis as db
    - Note: This is a workaround for deploying the bot on [railway.app](), will be dropped in the future
- Support arm64 (docker build)
- Support resending a message using a media relay server if Telegram cannot send a message with media due to Telegram
  server instability or network instability between media server and Telegram server
- Support colored logging
- `docker-compose.yml.sample`
- `/version` command to check bot version
- Automatically use proxy if global proxy (env var `SOCKS_PROXY`/`HTTP_PROXY`) set

#### Changes

- Assign feed monitoring tasks to every minute, instead of executing all at once each `DELAY`
    - Thus, env var `DELAY` can only be 60~3600
    - Note: env var `DELAY` will be deprecated in the future
- Recognize a post by its `guid`/`id` instead of `link`
- Simplify the output of `/list`
- Bump Python to 3.9 (docker build)
- Minor fixes

### zh-Hans

#### 重大变更

- 与 Telegram 交互的库由使用 HTTP Bot API 的同步库 `python-telegram-bot` 改为使用 MTProto Bot API 的异步库 `telethon`
    - 这引入了 API key 的需求，程序已经内置了 7 个公开的 API key，通常情况下不应无法登入。如果无法登入，可以自己申请 API key (
      详见 [docker-compose.yml.sample](https://github.com/Rongronggg9/RSS-to-Telegram-Bot/blob/53f11a473933e620d707c9d15f6d48737bd7a982/docker-compose.yml.sample#L43)
      中的说明)

#### 新特性

- 由于 Telegram bot 库的替换，bot 可以直接连接到 bot 所属的 DC，不需绕经 HTTP Bot API；也不需轮询获得消息更新，它在接收及发送消息方面都更为迅速，资源占用也更低； 即使 HTTP Bot API
  宕机，bot 也可以正常工作 (
  详见 [Advantages of MTProto over Bot API](https://docs.telethon.dev/en/latest/concepts/botapi-vs-mtproto.html#advantages-of-mtproto-over-bot-api)
  和 [MTProto vs HTTP Bot API](https://github.com/LonamiWebs/Telethon/wiki/MTProto-vs-HTTP-Bot-API))
- 支持更多元素的解析
    - `<iframe>`
    - `<video><source><source>...</video>`
    - `<code>`
    - `<pre>`
- 支持 OPML 导入导出
- 支持超长文章通过 Telegraph 发送 (必须先设置 `TELEGRAPH_TOKEN` 环境变量)
- 支持使用 redis 作为数据库
    - 注意：这是为了在 [railway.app]() 上部署而设计的变通解决方案，未来很可能丢弃
- 支持 arm64 (docker 构建)
- 支持在由于 Telegram 服务器不稳定或 Telegram 服务器与媒体服务器之间的网络连接不稳定而导致 Telegram 无法发出带有媒体的消息时，使用媒体反代服务器重新发送。
- 支持日志着色
- `docker-compose.yml.sample`
- 用于检查 bot 版本的 `/version` 命令
- 如果设置了全局代理 (环境变量 `SOCKS_PROXY`/`HTTP_PROXY`)，会使用它们

#### 变更

- 将 feed 监视任务分配到每分钟，而不是每次 `DELAY` 一次性全部执行
    - 因此，环境变量 `DELAY` 将只能被设置为 60~3600
    - 注意：环境变量 `DELAY` 未来将被弃用
- 使用 `guid`/`id` 来辨识一个 post，而不是 `link`
- 简化了 `/list` 的输出
- 升级为 Python 3.9 (docker 构建)
- 次要的修复

## v1.5.0

### en

- Post parser is completely rewritten, more stable and can keep text formatting as much as possible
- GIF Support
- When the message is more than 10 pieces of media, send it in pieces
- Support video and pictures to be mixed in the same message arbitrarily
- Invalid media are no longer directly discarded, but attached to the end of the message as a link
- Automatically determine whether the title of the RSS feed is auto-filled, if so, omit the title
- Automatically show the author name
- Automatically replace emoji shortcodes with emoji
- Automatically replace emoji images with emoji or its description text
- When an image cannot be sent due to the instability of telegram api, the image server will be automatically replaced
  and resent
    - Only for Weibo images, non-Weibo images will be attached to the end of the message as a link

src/env.py
#  RSS to Telegram Bot
#  Copyright (C) 2020-2025  Rongrong <i@rong.moe>
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU Affero General Public License as
#  published by the Free Software Foundation, either version 3 of the
#  License, or (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU Affero General Public License for more details.
#
#  You should have received a copy of the GNU Affero General Public License
#  along with this program.  If not, see <https://www.gnu.org/licenses/>.

from __future__ import annotations
from typing import Optional, Union
from typing_extensions import Final

import asyncio
import yarl
import os
import sys
import colorlog
import re
import argparse
from contextlib import suppress
from telethon import TelegramClient
from telethon.tl.types import User, InputPeerUser
from python_socks import parse_proxy_url
from dotenv import load_dotenv
from pathlib import Path
from functools import partial

from .version import __version__


# ----- utils -----
def __bool_parser(var: Optional[str], default_value: bool = False) -> bool:
    if not var:
        return default_value

    if var.isdecimal() or var.lstrip('-').isdecimal():
        return int(var) > 0

    var = var.upper()
    if var in ('FALSE', 'NONE', 'NULL', 'NO', 'NOT', 'DISABLE', 'DISABLED', 'INACTIVE', 'DEACTIVATED', 'OFF'):
        return False
    if var in ('TRUE', 'YES', 'OK', 'ENABLE', 'ENABLED', 'ACTIVE', 'ACTIVATED', 'ON'):
        return True
    return default_value


def __list_parser(var: Optional[str]) -> list[str]:
    return re.split(r'[\s,;，；]+', var.strip()) if var else []


def __decompose_version(version: str) -> list[Union[int, str]]:
    def buf_put(as_int: bool):
        nonlocal buf
        if not buf:
            raise ValueError(f'Empty part in version string: {version}')
        parts.append(int(buf) if as_int else buf)
        buf = ''

    parts = []
    buf = ''
    for i, char in enumerate(version):
        if char == '.':
            buf_put(as_int=True)
            continue
        if not char.isdecimal():
            buf_put(as_int=True)
            parts.append(version[i:])
            break
        buf += char
    if buf:
        buf_put(as_int=True)
    return parts


def __compare_version(version1: str, version2: str) -> tuple[int, list[Union[int, str]], list[Union[int, str]]]:
    """loose comparison, only compare the numeric parts"""
    parts1 = __decompose_version(version1)
    parts2 = __decompose_version(version2)

    if not parts1:
        return -1, parts1, parts2
    if not parts2:
        return 1, parts1, parts2

    marker = 0
    for part1, part2 in zip(parts1, parts2):
        if part1 == part2:
            continue
        part1_is_int = isinstance(part1, int)
        part2_is_int = isinstance(part2, int)
        if part1_is_int and part2_is_int:
            marker = 1 if part1 > part2 else -1
        elif part1_is_int and part1 > 0:
            marker = 1
        elif part2_is_int and part2 > 0:
            marker = -1
        break
    return marker, parts1, parts2


# ----- setup logging -----
__configure_logging = partial(
    colorlog.basicConfig,
    format='%(log_color)s%(asctime)s:%(levelname)s:%(name)s - %(message)s',
    datefmt='%Y-%m-%d-%H:%M:%S'
)
__configure_logging(level=colorlog.DEBUG if __bool_parser(os.environ.get('DEBUG')) else colorlog.INFO)
logger = colorlog.getLogger('RSStT.env')

# ----- determine the environment -----
user_home = os.path.expanduser('~')
self_path = os.path.dirname(__file__)
self_module_name = os.path.basename(self_path)
cli_entry = sys.argv[0]  # expect `-m` or `/path/to/telegramRSSbot.py`
is_self_run_as_a_whole_package = cli_entry.endswith('telegramRSSbot.py')

__arg_parser = argparse.ArgumentParser(
    prog=cli_entry if cli_entry != '-m' else f'python3 -m {self_module_name}',
    description='RSS to Telegram Bot, a Telegram RSS bot that cares about your reading experience.')
__arg_parser.add_argument('-c', '--config', metavar='/path/to/config/folder', type=str, nargs=1,
                          help='path to the config folder')
__arg_parser.add_argument('--dummy', type=str, action='append', nargs=1,
                          help='dummy argument, can be repeated for multiple times. '
                               'Useful to distinguish multiple bot instances on the same host with command line')
cli_args = __arg_parser.parse_args()
custom_config_path = cli_args.config[0] if cli_args.config else None

if custom_config_path:
    config_folder_path = os.path.normpath(os.path.abspath(custom_config_path))
elif is_self_run_as_a_whole_package:
    config_folder_path = os.path.normpath(os.path.join(self_path, '..', 'config'))
else:
    config_folder_path = os.path.join(user_home, '.rsstt')

Path(config_folder_path).mkdir(parents=True, exist_ok=True)
logger.info(f'Config folder: {config_folder_path}')

# ----- load .env -----
dot_env_paths = (os.path.join(config_folder_path, '.env'),
                 os.path.join(os.path.abspath('.'), '.env'))
if is_self_run_as_a_whole_package:
    dot_env_paths = (os.path.normpath(os.path.join(self_path, '..', '.env')),) + dot_env_paths
for dot_env_path in dict.fromkeys(dot_env_paths):  # remove duplicates while keeping the order
    if os.path.isfile(dot_env_path):
        load_dotenv(dot_env_path, override=True)
        logger.info(f'Found .env file at "{dot_env_path}", loaded')


# ----- get version -----
def __get_version():
    version = 'dirty'
    if is_self_run_as_a_whole_package:
        # noinspection PyBroadException
        try:
            with open(os.path.normpath(os.path.join(self_path, '..', '.version')), 'r') as v:
                version = v.read().strip()
        except Exception:
            version = 'dirty'

        if not version or version == '@':
            version = 'dirty'

    if version == 'dirty':
        import subprocess

        with suppress(Exception):
            # vMAJ.MIN.PAT-N-gHASH[-(broken|dirty)][@BRANCH]
            version = '@'.join(
                subprocess.run(
                    args,
                    cwd=self_path,
                    stdout=subprocess.PIPE,
                    stderr=subprocess.DEVNULL,
                    text=True,
                    check=True,
                    timeout=3,
                ).stdout.strip()
                for args in (
                    ['git', 'describe', '--tags', '--long', '--dirty', '--broken', '--always'],
                    ['git', 'branch', '--show-current'],
                )
            )

    try:
        sign, version_decomposed, version_pkg_decomposed = __compare_version(version.lstrip('v'), __version__)
        if sign <= -1:  # outdated version: older than pkg ver
            version = __version__
            if isinstance(version_decomposed[-1], str) and not isinstance(version_pkg_decomposed[-1], str):
                # -N-gHASH[-(broken|dirty)][@BRANCH]
                # ^^
                version += re.sub(r'^-\d+(?=-)', '', version_decomposed[-1])  # trim the marked part
            version = f'v{version}'
    except (TypeError, ValueError):
        version = f'v{__version__}'

    # empty guard
    if not version:
        version = 'dirty'

    return version


VERSION: Final = __get_version()

# ----- basic config -----
SAMPLE_APIS: Final = {
    # https://github.com/DrKLO/Telegram/blob/master/TMessagesProj/src/main/java/org/telegram/messenger/BuildVars.java
    4: '014b35b6184100b085b0d0572f9b5103',
    # https://github.com/TelegramMessenger/Telegram-iOS/blob/master/build-system/verify.sh
    8: '7245de8e747a0d6fbe11f7cc14fcc0bb',
    # https://github.com/overtake/TelegramSwift/blob/master/Telegram-Mac/Config.swift
    9: '3975f648bb682ee889f35483bc618d1c',
    # https://github.com/vysheng/tg/blob/master/loop.h
    2899: '36722c72256a24c1225de00eb6a1ca74',
    # https://github.com/telegramdesktop/tdesktop/blob/dev/docs/api_credentials.md
    17349: '344583e45741c457fe1862106095a5eb',
    # https://github.com/tdlib/td/blob/master/example/uwp/app/MainPage.xaml.cs
    94575: 'a3406de8d171bb422bb6ddf3bbd800e2',
    # https://github.com/morethanwords/tweb/blob/master/t.py
    1025907: '452b0359b988148995f22ff0f4229750'
}

API_ID: Final = int(os.environ['API_ID']) if os.environ.get('API_ID') else None
API_HASH: Final = os.environ.get('API_HASH')
TOKEN: Final = os.environ.get('TOKEN')

try:
    _manager = tuple(map(int, __list_parser(os.environ.get('MANAGER') or os.environ.get('CHATID'))))
    MANAGER: Final = set(_manager)  # optimize membership test

    if not all((TOKEN, MANAGER)):
        logger.critical('"TOKEN" OR "MANAGER" NOT SET! PLEASE CHECK YOUR SETTINGS!')
        exit(1)
except Exception as e:
    logger.critical('INVALID "MANAGER"! PLEASE CHECK YOUR SETTINGS!', exc_info=e)
    exit(1)

_error_logging_chat = os.environ.get('ERROR_LOGGING_CHAT')
ERROR_LOGGING_CHAT: Final = int(_error_logging_chat) if _error_logging_chat else _manager[0]
del _error_logging_chat
del _manager

MANAGER_PRIVILEGED: Final = __bool_parser(os.environ.get('MANAGER_PRIVILEGED'))

TELEGRAPH_TOKEN: Final = __list_parser(os.environ.get('TELEGRAPH_TOKEN'))

MULTIUSER: Final = __bool_parser(os.environ.get('MULTIUSER'), default_value=True)

CRON_SECOND: Final = int(os.environ.get('CRON_SECOND') or 0) % 60

# ----- network config -----
DEFAULT_PROXY: Final = os.environ.get('SOCKS_PROXY') or os.environ.get('socks_proxy') \
                       or os.environ.get('HTTP_PROXY') or os.environ.get('http_proxy') \
                       or os.environ.get('HTTPS_PROXY') or os.environ.get('https_proxy')

TELEGRAM_PROXY: Final = os.environ.get('T_PROXY') or DEFAULT_PROXY
if TELEGRAM_PROXY:
    _parsed = parse_proxy_url(TELEGRAM_PROXY.replace('socks5h', 'socks5'))
    TELEGRAM_PROXY_DICT: Final = {
        'proxy_type': _parsed[0],
        'addr': _parsed[1],
        'port': _parsed[2],
        'username': _parsed[3],
        'password': _parsed[4],
        'rdns': True
    }
    TELEGRAPH_PROXY_DICT: Final = {
        'proxy_type': _parsed[0],
        'host': _parsed[1],
        'port': _parsed[2],
        'username': _parsed[3],
        'password': _parsed[4],
        'rdns': True
    }
    del _parsed
else:
    TELEGRAM_PROXY_DICT: Final = None
    TELEGRAPH_PROXY_DICT: Final = None

R_PROXY: Final = os.environ.get('R_PROXY') or DEFAULT_PROXY

REQUESTS_PROXIES: Final = {'all': R_PROXY} if R_PROXY else {}

PROXY_BYPASS_PRIVATE: Final = __bool_parser(os.environ.get('PROXY_BYPASS_PRIVATE'))
PROXY_BYPASS_DOMAINS: Final = __list_parser(os.environ.get('PROXY_BYPASS_DOMAINS'))
USER_AGENT: Final = os.environ.get('USER_AGENT') or f'RSStT/{__version__} RSS Reader (+https://git.io/RSStT)'
IPV6_PRIOR: Final = __bool_parser(os.environ.get('IPV6_PRIOR'))
VERIFY_TLS: Final = __bool_parser(os.environ.get('VERIFY_TLS'), default_value=True)

HTTP_TIMEOUT: Final = int(os.environ.get('HTTP_TIMEOUT') or 12)
HTTP_CONCURRENCY: Final = int(os.environ.get('HTTP_CONCURRENCY') or 1024)
HTTP_CONCURRENCY_PER_HOST: Final = int(os.environ.get('HTTP_CONCURRENCY_PER_HOST') or 16)
HTTP_MAX_LINE_SIZE: Final = int(os.environ.get('HTTP_MAX_LINE_SIZE') or 16384)
HTTP_MAX_FIELD_SIZE: Final = int(os.environ.get('HTTP_MAX_FIELD_SIZE') or 16384)
DNS_CACHE_TTL: Final = int(os.environ.get('DNS_CACHE_TTL') or 10)

# ----- img relay server config -----
_img_relay_server = os.environ.get('IMG_RELAY_SERVER') or 'https://rsstt-img-relay.rongrong.workers.dev/'
IMG_RELAY_SERVER: Final = (
        ('' if _img_relay_server.startswith('http') else 'https://')
        + _img_relay_server
        + ('' if _img_relay_server.endswith(('/', '=')) else '/')
)
del _img_relay_server

# ----- wsrv.nl config -----
_images_weserv_nl = os.environ.get('IMAGES_WESERV_NL') or 'https://wsrv.nl/'
IMAGES_WESERV_NL: Final = (
        ('' if _images_weserv_nl.startswith('http') else 'https://')
        + _images_weserv_nl
        + ('' if _images_weserv_nl.endswith('/') else '/')
)
del _images_weserv_nl


# ----- db config -----
def __get_database_url() -> str:
    # Railway.app provides DATABASE_PRIVATE_URL, DATABASE_URL and/or DATABASE_PUBLIC_URL.
    # DATABASE_PRIVATE_URL is for private networking, while DATABASE_PUBLIC_URL is for public networking.
    # If private networking is disabled, the former still exists but is invalid (i.e., missing host).
    # If public networking is disabled, both host and port are missing in the latter, resulting in an invalid URL too.
    # A legacy Postgres addon may provide only DATABASE_URL (for PUBLIC networking) and DATABASE_PRIVATE_URL,
    # while a modern one may provide only DATABASE_URL (for PRIVATE networking) and DATABASE_PUBLIC_URL.
    # Thus, we need to select a valid one from them.
    urls: tuple[str, ...] = tuple(filter(None, map(os.environ.get, (
        'DATABASE_URL',  # Generic, Railway.app compatible
        'DATABASE_PRIVATE_URL',  # Railway.app specific
        'DATABASE_PUBLIC_URL',  # Railway.app specific
    ))))
    if not urls:
        return f'sqlite:{config_folder_path}/db.sqlite3'
    err: Optional[BaseException] = None
    for url in urls:
        try:
            y_url = yarl.URL(url)
        except ValueError as _err:
            err = _err
        else:
            return str(
                # Tortoise-ORM does not recognize 'postgresql' scheme
                y_url.with_scheme('postgres')
                if y_url.scheme == 'postgresql'
                else y_url
            )
    else:
        logger.critical('INVALID DATABASE URL!', exc_info=err)


DATABASE_URL: Final = __get_database_url()

# ----- misc config -----
TABLE_TO_IMAGE: Final = __bool_parser(os.environ.get('TABLE_TO_IMAGE'))
TRAFFIC_SAVING: Final = __bool_parser(os.environ.get('TRAFFIC_SAVING'))
LAZY_MEDIA_VALIDATION: Final = __bool_parser(os.environ.get('LAZY_MEDIA_VALIDATION'))
NO_UVLOOP: Final = __bool_parser(os.environ.get('NO_UVLOOP'))
MULTIPROCESSING: Final = __bool_parser(os.environ.get('MULTIPROCESSING'))
EXECUTOR_NICENESS_INCREMENT: Final = int(os.environ.get('EXECUTOR_NICENESS_INCREMENT') or (2 * hasattr(os, 'nice')))
DEBUG: Final = __bool_parser(os.environ.get('DEBUG'))
__configure_logging(  # config twice to make .env file work
    level=colorlog.DEBUG if DEBUG else colorlog.INFO,
    force=True
)
if DEBUG:
    logger.debug('DEBUG mode enabled')

# ----- environment config -----
RAILWAY_STATIC_URL: Final = os.environ.get('RAILWAY_STATIC_URL')
PORT: Final = int(os.environ.get('PORT', 0)) or (8080 if RAILWAY_STATIC_URL else None)

# !!!!! DEPRECATED WARNING !!!!!
if os.environ.get('DELAY'):
    logger.warning('Env var "DELAY" is DEPRECATED and of no use!\n'
                   'To avoid this warning, remove this env var.')

if os.environ.get('CHATID'):
    logger.warning('Env var "CHATID" is DEPRECATED!\n'
                   'To avoid this warning, remove this env var.')

if any((os.environ.get('REDISHOST'), os.environ.get('REDISUSER'), os.environ.get('REDISPASSWORD'),
        os.environ.get('REDISPORT'), os.environ.get('REDIS_NUM'),)):
    logger.warning('Redis DB is DEPRECATED!\n'
                   'ALL SUBS IN THE OLD DB WILL NOT BE MIGRATED. '
                   'IF YOU NEED TO BACKUP YOUR SUBS, DOWNGRADE AND USE "/export" COMMAND TO BACKUP.\n\n'
                   'Please remove these env vars (if exist):\n'
                   'REDISHOST\n'
                   'REDISUSER\n'
                   'REDISPASSWORD\n'
                   'REDISPORT\n'
                   'REDIS_NUM')

if is_self_run_as_a_whole_package and os.path.exists(os.path.join(config_folder_path, 'rss.db')):
    os.rename(os.path.join(config_folder_path, 'rss.db'), os.path.join(config_folder_path, 'rss.db.bak'))
    logger.warning('Sqlite DB "rss.db" with old schema is DEPRECATED and renamed to "rss.db.bak" automatically!\n'
                   'ALL SUBS IN THE OLD DB WILL NOT BE MIGRATED. '
                   'IF YOU NEED TO BACKUP YOUR SUBS, DOWNGRADE AND USE "/export" COMMAND TO BACKUP.')

# ----- shared var -----
bot: Optional[TelegramClient] = None  # placeholder
bot_id: Optional[int] = None  # placeholder
bot_peer: Optional[User] = None  # placeholder
bot_input_peer: Optional[InputPeerUser] = None  # placeholder

# ----- loop initialization -----
uvloop_enabled = False
if not NO_UVLOOP:
    try:
        import uvloop

        uvloop.install()
        uvloop_enabled = True
    except ImportError:  # not installed (e.g. Windows)
        uvloop = None
loop = asyncio.new_event_loop()
asyncio.set_event_loop(loop)
